File: cpp/Value.cpp
#include "Value.h"
#include "Object.h"

std::string Value::Stringify() const
{
    switch (type)
    {
    case ValueType::NIL:
        return "nil";
    case ValueType::NUM:
        return std::to_string(stored);
    case ValueType::BOOL:
        return stored == 1.0 ? "true" : "false";
    case ValueType::OBJECT:
        return ObjectStringify(object);
    default:
        return "nil";
    }
    return "nil";
}

void Value::Mark() const
{
    if (type == ValueType::OBJECT)
        ObjectMark(object);

}
void Value::UnMark() const
{
    if (type == ValueType::OBJECT)
        ObjectUnMark(object);
}

bool Value::IsValid() const
{
    return (type<=ValueType::OBJECT && type>=ValueType::NUM) && object;
}

bool operator==(const Value &left, const Value &right)
{
    if (left.type != right.type)
        return false;

    switch (left.type)
    {
    case ValueType::NIL:
        return IS_NIL_VALUE(right);
    case ValueType::NUM:
        return left.stored == TO_NUM_VALUE(right);
    case ValueType::BOOL:
        return left.stored == TO_BOOL_VALUE(right);
    case ValueType::OBJECT:
        return IsObjectsEqual(left.object, TO_OBJECT_VALUE(right));
    default:
        return false;
    }
}

bool operator!=(const Value &left, const Value &right)
{
    return !(left == right);
}


Value FindActualValue(const Value &v)
{
    auto value = GetEndOfRefValue(v);
    if (IS_BUILTIN_VALUE(value) && TO_BUILTIN_VALUE(value)->Is<Value>())
        value = TO_BUILTIN_VALUE(value)->Get<Value>();
    return value;
}

Value *GetEndOfRefValue(Value *v)
{
    auto result = v;
    while (IS_REF_VALUE(*result))
        result = TO_REF_VALUE(*result)->pointer;
    return result;
}

Value GetEndOfRefValue(const Value &v)
{
    auto value = v;
    while (IS_REF_VALUE(value))
        value = *TO_REF_VALUE(value)->pointer;
    return value;
}

File: c#/Value.cs
namespace ComputeDuck
{
    public enum ValueType
    {
        NIL,
        NUM,
        BOOL,
        OBJECT,
        UNKNOWN
    }

    public class Value
    {
        public Value()
        {
            this.type = ValueType.NIL;
        }

        public Value(double number)
        {
            this.number = number;
            this.type = ValueType.NUM;
        }

        public Value(bool boolean)
        {
            this.boolean = boolean;
            this.type = ValueType.BOOL;
        }

        public Value(Object obj)
        {
            this.obj = obj;
            this.type = ValueType.OBJECT;
        }

        public Value(ValueType type)
        {
            this.type = type;
        }

        public ValueType Type()
        {
            return type;
        }

        public string Stringify()
        {
            switch (type)
            {
                case ValueType.NIL:
                    return "nil";
                case ValueType.NUM:
                    return number.ToString();
                case ValueType.BOOL:
                    return boolean ? "true" : "false";
                case ValueType.OBJECT:
                    return obj.Stringify();
                default:
                    return "nil";
            }
        }

        public bool IsEqualTo(Value other)
        {
            switch (type)
            {
                case ValueType.NIL:
                    return other.Type() == ValueType.NIL;
                case ValueType.UNKNOWN:
                    {
                        if (other.type == ValueType.UNKNOWN)
                            return true;
                        return false;
                    }
                case ValueType.NUM:
                    {
                        if (other.Type() == ValueType.NUM)
                            return number == other.number;
                        else
                            return false;
                    }
                case ValueType.BOOL:
                    {
                        if (other.Type() == ValueType.BOOL)
                            return boolean == other.boolean;
                        return false;
                    }
                case ValueType.OBJECT:
                    {
                        if (other.Type() == ValueType.OBJECT)
                            return obj.IsEqualTo(other.obj);
                        return false;
                    }
                default:
                    return other.Type() == ValueType.NIL;
            }
        }

        public ValueType type;

        public double number;
        public bool boolean;
        public Object obj;
        public static readonly Value g_UnknownValue = new Value(ValueType.UNKNOWN);
    }

}